### PR TITLE
Fix: Message can't be encrypted with ethereum web3 provider

### PR DESCRIPTION
**Pull request by Rgascoin**
Related to: https://github.com/aleph-im/aleph-sdk-ts/issues/114

Working on implementation specifically [for metamask](https://docs.metamask.io/guide/rpc-api.html#unrestricted-methods).
As this feature was already implemented in the SDK, I think finishing the implementation can make sense here.

Besides this fix, this PR will also add a new feature to allow encryption testing inside the Toolshed. 

Be aware of the potential [security issue](https://medium.com/metamask/metamask-api-method-deprecation-2b0564a84686) using this method.

## Breaking change

This PR introduce the decrypt function inside ECIESAccount:
`abstract decrypt(content: Buffer): Promise<Buffer>;`

This means the NULS2 `decrypt() `function needs to wait from now.

### CODE REVIEW

**Comment by BjrInt** [comment]
Getting this error in the Toolshed app when trying to decrypt content with an Ethereum (and also oddly enough with a NULS2 account). Does not matter if this is instanciated from Metamask or from a Keypair (did not test it using a Ledger device yet).
image

**Review by BjrInt — CHANGES_REQUESTED** [review]

File: src/accounts/nuls2.ts [hunk 1]
```diff
@@ -60,8 +60,10 @@ export class NULS2Account extends ECIESAccount {
      *
      * @param encryptedContent The encrypted content to decrypt.
      */
-    decrypt(encryptedContent: Buffer): Buffer {
-        return secp256k1_decrypt(this.privateKey, encryptedContent);
+    decrypt(encryptedContent: Buffer): Promise<Buffer> {
+        return new Promise((resolve) => {
+            resolve(secp256k1_decrypt(this.privateKey, encryptedContent));
+        });
```
[the remark this answers is not in the document]
Comment (Rgascoin) [reply src/accounts/nuls2.ts]: Thanks, Fixed in: [5bd3d79](https://github.com/aleph-im/aleph-sdk-ts/pull/116/commits/5bd3d7969f09ec7bbca5e656740246314bbae7c0)

File: src/accounts/avalanche.ts [hunk 2]
```diff
@@ -38,14 +41,35 @@ export class AvalancheAccount extends ECIESAccount {
...
             publicKey = delegateSupport instanceof ECIESAccount ? delegateSupport.publicKey : delegateSupport;
         else publicKey = this.publicKey;
-        if (publicKey) return secp256k1_encrypt(publicKey, content);
 
-        throw new Error("Cannot encrypt content");
+        if (publicKey && !this.provider?.isMetamask()) {
+            // Wallet encryption method or non-metamask provider
+            return new Promise((resolve) => {
+                resolve(secp256k1_encrypt(publicKey, content));
+            });
```
[the remark this answers is not in the document]
Comment (Rgascoin) [reply src/accounts/avalanche.ts]: Thanks, Fixed in: [5bd3d79](https://github.com/aleph-im/aleph-sdk-ts/pull/116/commits/5bd3d7969f09ec7bbca5e656740246314bbae7c0)

File: src/accounts/ethereum.ts [hunk 3]
```diff
@@ -35,16 +38,33 @@ export class ETHAccount extends ECIESAccount {
...
+                                publicKey: publicKey,
+                                data: content.toString(),
+                                version: "x25519-xsalsa20-poly1305",
+                            }),
+                        ),
+                        "utf-8",
+                    ),
+                );
+            });
+        }
```
[the remark this answers is not in the document]
Comment (Rgascoin) [reply src/accounts/ethereum.ts]: Thanks, Fixed in: [5bd3d79](https://github.com/aleph-im/aleph-sdk-ts/pull/116/commits/5bd3d7969f09ec7bbca5e656740246314bbae7c0)

File: src/accounts/providers/JsonRPCWallet.ts [hunk 4]
```diff
@@ -82,9 +108,17 @@ export class JsonRPCWallet extends BaseProviderWallet {
...
+                await this.provider.send("wallet_switchEthereumChain", [{ chainId: "0x1" }]);
+            } else await this.provider.send("wallet_addEthereumChain", [ChainData[chainOrRpc]]);
+        } else {
+            await this.provider.send("wallet_addEthereumChain", [chainOrRpc]);
+        }
+    }
+
+    public isMetamask(): boolean {
+        return !!this.provider?.provider?.isMetaMask;
     }
```
[the remark this answers is not in the document]
Comment (Rgascoin) [reply src/accounts/providers/JsonRPCWallet.ts:123]: This function is also used for the encryption to detect if the metamask standards encryption is needed

**Review by BjrInt — APPROVED** [review]
LGTM

**Pull request merged**